ui/client/src/dialogs/ABIUpload.tsx

```diff
@@ -0,0 +1,203 @@
+// Copyright © 2024 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  Alert,
+  Box,
+  Button,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+  FormControlLabel,
+  Grid2,
+  Radio,
+  RadioGroup,
+  TextField,
+  Typography
+} from '@mui/material';
+import { FileUploader } from 'react-drag-drop-files';
+import { Dispatch, SetStateAction, useEffect, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { UploadFile } from '@mui/icons-material';
+import { useMutation } from '@tanstack/react-query';
+import { uploadABI } from '../queries/storeABI';
+
+type Props = {
+  dialogOpen: boolean
+  setDialogOpen: Dispatch<SetStateAction<boolean>>
+}
+
+export const ABIUploadDialog: React.FC<Props> = ({
+  dialogOpen,
+  setDialogOpen
+}) => {
+
+  const { t } = useTranslation();
+  const [errorMessage, setErrorMessage] = useState<string>();
+  const [fileSelected, setFileSelected] = useState<File | null>(null);
+  const [radioSelection, setRadioSelection] = useState<'file' | 'text'>('file');
+  const [abiText, setAbiText] = useState('');
+  const [abiUploadCount, setAbiUploadCount] = useState(0);
+
+  const { mutate, data, reset } = useMutation({
+    mutationFn: (value: Object) => uploadABI(value)
+  });
+
+  useEffect(() => {
+    if (dialogOpen) {
+      reset();
+      setRadioSelection('file');
+      setFileSelected(null);
+      setAbiText('');
+    }
+  }, [dialogOpen]);
+
+  const handleSubmit = async () => {
```

Comment:
See final review comment for more on this section

Reply:
This is a very interesting idea. I like being able to validate the ABI on the client before it is submitted to the Paladin node. @dechdev provided another interesting source for getting a comprehensive Zod based validator for ABI (https://github.com/wevm/abitype/tree/main/packages/abitype) which we could use. I started experimenting with it and built a prototype. It works exactly as expected: valid ABIs pass, and invalid throw errors, but exploring the error details can lead to long complex structures. Here is an example of a small ABI (simple storage) to which I introduced 1 small typo (input -> nput)

```json
ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": "constructor",
            "code": "invalid_literal",
            "expected": "error",
            "path": [
              0,
              "type"
            ],
            "message": "Invalid literal value, expected \"error\""
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              0,
              "inputs"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              0,
              "name"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "constructor",
            "code": "invalid_literal",
            "expected": "event",
            "path": [
              0,
              "type"
            ],
            "message": "Invalid literal value, expected \"event\""
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              0,
              "inputs"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              0,
              "name"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              0,
              "inputs"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      0
    ],
    "message": "Invalid input"
  }
]
```

It looks like if there is an error present, and its location is relatively hight in the object hierarchy within the ABI, that can cascase into many secondary errors.

I think this functionaltiy merits its own separate issue where we can investigate more in detail how to get the most out of Zod in order to improve user experience when it comes to raising problems in ABIs.

Reply:
For error validation, this is the recommended package from the Zod team: https://github.com/causaly/zod-validation-error

Reference: https://zod.dev/?id=error-handling

Reply:
Right, that's the approach I was using. In the examples that are shown in the Zod documentation, validation is clear for the simple flat-objects used. However, When using an Abi combined with the schema from AbiType, the errors that Zod list for a single type can be difficult to understand:

<img width="1231" alt="image" src="https://github.com/user-attachments/assets/886493fc-68b2-48bd-a4ab-5ef0bf3ac14f" />